### ISSUE TITLE
Audio pop/click after intro in Monkey Island 1 CD version using BIN+CUE

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#401]
Reported by @flashmasta, a click/pop can be heard when pressing `esc` after the intro in Monkey Island 1 using a single BIN/CUE. Other data points:

- When inspecting the raw sample values in the BIN file, the lead into to the audio tracks is clean (zeros).
- The presence of the click/pop appears to vary based on platform (cannot be heard on Linux).
- This click/pop doesn't happen when playing the same track using DOS-based CD-DA players.

Prior discussion in:
- https://github.com/dosbox-staging/dosbox-staging/issues/385
- https://github.com/dosbox-staging/dosbox-staging/pull/393

**Comment by flashmasta** [comment]
Example: https://drive.google.com/file/d/13PHJo-wEaJpCDomcvrTxfLWPVDZyl86P/view?usp=sharing

The popping when the audio first starts before the Lucasfilm logo (at about 00:04) is the loudest and most jarring.  There is another tiny pop @ 00:07 right before the CD track starts.  At 00:12 I press ESC to kill the logo screen, 00:15 is a tiny pop before the CD track for the cliffside scene starts, 00:16 I press ESC, 00:17 there is a tiny pop before the CD track starts for the "Part One" screen.

**Comment by kcgen** [comment]
Thanks @flashmasta.

Comparing your Windows (top) recording to Linux (bottom):

![Screenshot at 2020-05-24 14-39-01](https://user-images.githubusercontent.com/1557255/82765636-5ea5c480-9dcd-11ea-9399-dda525c8fe52.png)

![Screenshot at 2020-05-24 14-43-11](https://user-images.githubusercontent.com/1557255/82765627-5057a880-9dcd-11ea-8917-b32ae5f7d692.png)

Near the right in both graphs, they share what appears to be a single lone blip in the middle of the silence.  These blips are actually two samples up-close if we zoom in (windows then Linux):

![Screenshot at 2020-05-24 14-51-21](https://user-images.githubusercontent.com/1557255/82765812-c577ad80-9dce-11ea-8673-d9bddb0284de.png)

![Screenshot at 2020-05-24 14-53-55](https://user-images.githubusercontent.com/1557255/82765814-c577ad80-9dce-11ea-8d74-25e31b08a3aa.png)

What's unique in the Windows screenshot on the far-left is your initial negative blip.  
Up close, it's a 6-sample ramp dropping to nothing (I don't have an explanation for where this is coming from - can you try disabling all your sound card devices in your conf file?

![Screenshot at 2020-05-24 14-52-12](https://user-images.githubusercontent.com/1557255/82765813-c577ad80-9dce-11ea-847b-2cb1cbeec719.png)

**Comment by kcgen** [comment]
@flashmasta, @dreamer ,

Ok - I captured the actual values being handed over from the CD-DA player to dosbox's mixer on both the start and stop track events; and on start up it never passes junk values - it only passes the exact values decoded (or read) from the track file.

Likewise on shutdown, the very last sample handed over is from the file, and never hands over empty/uninitialized values.  So that's the good news - this problem exists outside the CD-DA code.

**Comment by kcgen** [comment]
The very first "pop" at the 4-second mark in your recording @flashmasta is coming from the sound blaster:

![](https://user-images.githubusercontent.com/1557255/82765813-c577ad80-9dce-11ea-847b-2cb1cbeec719.png)

Set `sbtype = none`, and this pop disappears. I was able to reproduce this on Linux.

**Comment by dreamer** [comment]
Is there a chance this problem is originating from Mixer? :( Does it happen with `sb1` or `sb2` or `gb` as well?

**Comment by kcgen** [comment]
The two-frames of "pop" between track changes are coming from dosbox's mixer.  

When a mixer channel is re-enabled, the channel (in the current code) may contain sample-values and tally-values at the moment when the channel was previously disabled.  

When the channel is enabled again, these values are processed ahead of the new inbound samples, and that's where the pop comes from.  The "pop" are left unprocessed samples the moment we pressed `esc` at the prior scene.

[This branch](https://github.com/dosbox-staging/dosbox-staging/compare/kc/mixer-pop-on-resume-1) makes the mixer's channel clear its sample value and tally counts when it goes dormant, so it starts up clean.

This fix is general, and all audio devices that disable and re-enable their mixer channel will benefit from a clean resume.

Windows build [is here](https://github.com/dosbox-staging/dosbox-staging/suites/713949460/artifacts/6991447), if you want to give it a go (with `sbtype = none` too :-)

**Comment by kcgen** [comment]
> Is there a chance this problem is originating from Mixer? :( Does it happen with sb1 or sb2 or gb as well?

Oh wow; you're right: it's only `sbtype = sb16` that generates that saw-tooth on startup. `sbpro2` (and lower) starts up clean.  Is this some 16-bit DMA init stuff that we can cleanup?

Here's some sound-blaster code that increases the shift size and sample position in a step-wise pattern:

```
			MixTemp[done++]=decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 6) & 0x3,sb.adpcm.reference,sb.adpcm.stepsize);
			MixTemp[done++]=decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 4) & 0x3,sb.adpcm.reference,sb.adpcm.stepsize);
			MixTemp[done++]=decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 2) & 0x3,sb.adpcm.reference,sb.adpcm.stepsize);
			MixTemp[done++]=decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 0) & 0x3,sb.adpcm.reference,sb.adpcm.stepsize);
```

**Comment by dreamer** [comment]
I really can't tell - this was shot in the dark based on the fact, that all older bugs we had with SoundBlaster emulation were not reproducible in `sb1`.

I looked through your commit and really like the changes in there - was the core of the issue caused by `MixerChannel::FillUp` locking SDL for a brief moment on start?

**Comment by kcgen** [comment]
> I really can't tell - this was shot in the dark based on the fact, that all older bugs we had with SoundBlaster emulation were not reproducible in sb1.

Geeze.. I can't even compile `DOSBox` anymore to see if this `sb16` issue also affects SVN. 

``` c++
./src/video/x11/SDL_x11events.c:1433:6: note: code may be misoptimized unless ‘-fno-strict-aliasing’ is used
./src/video/SDL_blit_A.c: Assembler messages:
./src/video/SDL_blit_A.c:2044: Error: operand type mismatch for `movq'
...
g++-9 returned 1 exit status
```
> I looked through your commit and really like the changes in there

Thanks! I'm pretty sure this will solve this; will see what @flashmasta reports.

> was the core of the issue caused by MixerChannel::FillUp locking SDL for a brief moment on start?

The crux is in `Enable()`.

When we press `esc` in MI1, the CD-DA immediately calls `channel->Enable(false)` and stops feeding it samples.  However, the mixer's channel leaves the existing sample values (prev_sample[] and next_sample[]) as-is at the time when we halted the channel (along with the `done` and `remaining` tallies). 

When MI1 changes tracks, CD-DA re-enables the channel however those tallies and prior samples are processed first - and that's the "pop". 

So it's important that `Enable(false)` scrub the channel back down to clean slate, just like on startup, when everything is zero'd the first time a channel is enabled.

**Comment by flashmasta** [comment]
I can't hear any clicks between tracks using the provided build above.  The only remain sound is the first click at game load when sbtype = sb16.  When sbtype = none or sbpro2 it is a clean start on my end.

**Comment by kcgen** [comment]
Roger that @flashmasta - thanks for the quick tests and confirmation!
Promoting this PR from draft to review.

**Comment by flashmasta** [comment]
FYI, I can reproduce the first pop issue on dosbox 74-3 when sbtype=sb16 every time, switching to none or sbpro2 will remove that initial pop every time as well.  Oddly enough dosbox-x will never produce that initial pop on first play when sbtype=sb16.  It isn't until you exit the game and load it a subsequent time that it will pop.

I'm not sure that helps but I thought the dosbox-x behavior was odd that it stands out from the rest.

**Comment by kcgen** [comment]
Nice digging @flashmasta ; if we can fix this, I suspect the DOSBox Team might be interested in accepting this upstream too. 

I added some code to catch the point when that ramp is fed to SDL; here's the backtrace:

```
MixerChannel::AddSamples<unsigned char, false, false, true> (this=0x55555a6f1fa0, len=1, data=0x555558733b00 <sb+64> "") at mixer.cpp:380
MixerChannel::AddSamples_m8 (this=0x55555a6f1fa0, len=1, data=0x555558733b00 <sb+64> "") at mixer.cpp:431
GenerateDMASound (size=1) at sblaster.cpp:489
END_DMA_Event (val=1) at sblaster.cpp:607
PIC_RunQueue () at pic.cpp:495
Normal_Loop () at dosbox.cpp:138
DOSBOX_RunMachine () at dosbox.cpp:322
```

**Comment by kcgen** [comment]
**@ripsaw8080** replied:

> A small DMA transfer is not unusual for SB detection / initialization when games start up. It can cause an audible pop depending on the sample(s) in the transfer. The difference between SB16 and other SB types is likely due to the speaker enable state. My use of the word "speaker" here refers to a SB DSP flag, not the PC speaker. A disabled speaker flag doesn't actually silence output on an SB16, but it does in older models. You can probably see this in DOSBox debugger log messages where the init transfer is logged as "Short transfer scheduling IRQ ..." for sbtype=sb16, and "Silent DMA Transfer scheduling IRQ..." for other sbtype= settings.

I asked if these could be safely discarded, under the following conditions: it's the first-ever DMA transfer and if the size is less-than-some-threshold, then discard it one-time-only.

He suggested: 

> I may have become relatively insensitive because there are games (usually not from big name developers, mind you) doing stuff during SB initialization that is much worse for the ears than a small pop. In any case, I can safely suggest not going too far with hacks.

So this is (unfortunately) normal hardware behavior for the SB16.  Indeed, I also recall plenty of games causing an initial pop on startup once I upgraded my hardware in '92 - 93.. it was practically a sign that your audio card was properly detected!

I think if we did want to try to eat this, it would make sense to survery the initial DMA transfer size from the set of [DOS 16-bit audio games](https://www.vogons.org/viewtopic.php?t=54670), so we could choose a safe lower-bound.

For the other 98% of DOS games that never did use 16-bit audio, we should simply use `sbpro2`, and ripsaw8080's explains that properly honored the `speaker enable state` state and stayed mute during these initialization events.

**Comment by flashmasta** [comment]
If it is normal hardware behavior for it to have that initial pop in certain instances then I would personally prefer to have it 100% accurate, warts and all.  You've already smoothed out the other 99% of the pops and clicks in MI1 and should clean up a lot of other titles as well I would think.

**Comment by kcgen** [comment]
@flashmasta, thanks for the kind words. 

Our ears [have a protective response](http://hyperphysics.phy-astr.gsu.edu/hbase/Sound/protect.html) when natural sounds get too loud, unfortunately, discontinuities (going from zero-to-something) instantaneously is not nature (even the clap of lightening and thunder has a ramp-up), and unfortunately our ears can't respond in time - which can result in hearing damage (especially for those wearing headphones).

If pixel-perfect scaling is a quality of life improvement to our eyes, seamless mouse integration a quality of life improvement to our hands, and `ls` and `tab-command` completion a quality of life improvement to our expectations, then eliminating Creative Labs' hearing-damage-inducing initial pop of the sound-blaster 16 is a quality of life improvement for our ears.

Creative Labs has used good marketing to foist decades of poor engineering on the world; if there's something we can do in the software-world to correct it (with probably three-lines of code), then that's a small win for users.  That said - this still remains to be seen! Creative Labs might still get the last laugh.

**Comment by kcgen** [comment]
:cough: well, after all that tough talk, I tested all but a couple of those [16-bit-audio games](**https://www.vogons.org/viewtopic.php?t=54670**) and I can't find one of them that generates a pop on startup like Monkey Island 1 does when using `sbtype = sb16`.

Then again, Monkey Island 1 isn't even on that list because it only uses 8-bit sound samples.. so my guess is there's probably quite a few similar 8-bit-audio games that also pop on startup with `sbtype = sb16` just like Mi1 (and similarly, they would be more accurately served using  `sbpro2`).

If we go by the numbers, the vast majority of DOS games only need `sbpro2`; so in theory this should be our default. On the other hand, defaulting to `sb16` has pros and cons: it covers both the 16-bit and 8-bit audio games, with the con being the risk of pop-on-startup for some number of games, just like Mi1.

**Comment by dreamer** [comment]
#404 Is now merged to master; I am not sure if we want to close this issue already or maybe continue discussion. @flashmasta what do you think?

@kcgen I was wondering if there's actually any difference between sound quality of both cards (if there was not then I would be ok with using `sbpro2` as default), but this comparison proves the difference is huge: https://www.youtube.com/watch?v=t0tFKB7bvlA

Also, this Pixelmusement video might be helpful: https://www.youtube.com/watch?v=jLeQ2ZAbr3U

Perhaps we should create a new label "SB16 emulation" and start marking specific bug reports? (this is the 3rd SB16-related bug, I believe)

As for first "pop" in SB16… when `sb16` is selected, we could propagate info about `speaker enable` to the DOSBox mixer and put very short compressor/filter to eliminate the problem. But it's only an idea, I don't know if it would work in practice; if we added only a simple boolean flag without propagating "enable" info from SB emulation to dosbox mixer, then the behaviour would match DOSBox-X (with pop 

Overall, on Linux, users can work around this by using [PulseEffects](https://github.com/wwmm/pulseeffects) and turning on their own compressors, but Windows users might not have easy access to such software… and it's not nice to force users into using additional software to work around problems in project we control.

**Comment by kcgen** [comment]
> Perhaps we should create a new label "SB16 emulation" and start marking specific bug reports? (this is the 3rd SB16-related bug, I believe)

That would be great - to have a list of issues to work again for each sound card.

> I was wondering if there's actually any difference between sound quality of both cards 

Both sb16 and sbpro2 handle rates of up to 44.1 kHz, so the difference comes down source-samples, and this is where _almost_ every DOS game passes 8-bit PCM samples to the sound-card using the low (8-bit) DMA transfers.  So for the vast majority of games, there really isn't a difference.

There's only 41 games discovered so far that pass 16-bit PCM samples to the sound card (those games listed in that vogons thread), and indeed, using `sb16` or `gus` will buy you the best quality for them.  But yeah, 41 games amounts to 0.58% versus the other 99.4% of games that use 8-bit samples (if you look at Exo's DOS-only archive).

For example, Jazz Jack Rabbit gives options for configuring SB16 vs SB Pro - and various "qualities" (I set both to "Pentium - Highest quality"). I then configured `sbtype` to match the configuration, and sure enough there is visible or audible difference in the quality.

![Screenshot at 2020-05-25 15-39-01](https://user-images.githubusercontent.com/1557255/82847133-3d160d00-9ea1-11ea-9acb-06c650f76e94.png)

[jazz-sb16_vs_sbpro2.zip](https://github.com/dosbox-staging/dosbox-staging/files/4679120/jazz-sb16_vs_sbpro2.zip)
